---
title: "Class 11: Structural Bioinformatics II/AlphaFold"
author: "Wade Ingersoll (PID: A69038080)"
format: pdf
---

## Background

We saw last day that the PDB has 209,886 entries (Oct/Nov 2025). UniProtKB (i.e. protein sequence database) has 199,579,901 entries.

```{r}
209886/199579901 * 100
```

So, the PDB has only 0.1% coverage of the main sequence database.

Enter AlphaFold database (AFDB). < https://alphafold.ebi.ac.uk/ > that attempts to provide computed models for all sequences in UniProt.

"AlphaFold DB provides open access to over 200 million protein structure predictions to accelerate scientific research."

## AlphaFold

AlphaFold has 3 main outputs/results:

- The predicted coordinates (PDB files)
- A local quality score called **pLDDT** (one for each amino acid)
- A second quality score **PAE** Predicted Aligned Error (for each pair of amino acids)


We can run AlphaFold ourselves if we are not happy with AFDB (i.e. no coverage or poor model).

## Interpreting/Analyzing AF results in R

```{r}
results_dir <- "HIVPR_dimer_23119/"
```

```{r}
# File names for all PDB models
pdb_files <- list.files(path=results_dir,
                        pattern="*.pdb",
                        full.names = TRUE)

# Print our PDB file names
basename(pdb_files)
```
```{r}
library(bio3d)
library(bio3d.view)

pdbs <- pdbaln(pdb_files, fit=TRUE, exefile="msa")
```
```{r}
view.pdb(pdbs)
```
```{r}
aln_file <- list.files(path=results_dir,
                       pattern=".a3m$",
                        full.names = TRUE)
aln_file
```

```{r}
sim <- conserv(aln_file)
```








