---
title: "Class 12: RNA-seq with DESeq2"
author: "Wade Ingersoll (PID: A69038080)"
format: pdf
toc: true
---

## Background

Today we will analyze some RNAseq data from Himes et al. on the effects of a common steroid (dexamethasone also called "dex") on airway smooth muscle cells (ASMs).

For this analysis we need two main inputs

-`countData`: a table of **counts** per gene (in rows) across experiments (in columns)

-`colData`: **metadata** about the design of the experiments. The rows here must match the columns in `countData`

## Data Import

```{r}
counts <- read.csv("/Users/wadeingersoll/Desktop/BGGN213/class12/airway_scaledcounts.csv", row.names = 1)
metadata <- read.csv("/Users/wadeingersoll/Desktop/BGGN213/class12/airway_metadata.csv")
```

```{r}
head(counts)
```

```{r}
metadata
```

> Q1: How many "genes" in this dataset? - 38,694

```{r}
nrow(counts)
```

> Q2: How many experiments (i.e. columns in `counts` or rows in `metadata`) are there? - 8

```{r}
ncol(counts)
```

> Q3: How many "control" experiments are in the dataset?

```{r}
sum(metadata$dex == "control")
```
## Toy analysis example

1. Extract the "control" columns from `counts`
2. Calculate the mean value for each gene in these "control" columns

3-4. Do the same for the "treated" columns
5. Compare these mean values for each gene


Step 1.
```{r}
control.inds <- metadata$dex == "control"
control.counts <- counts[ , control.inds]
```

Step 2.
```{r}
control.mean <- rowMeans(control.counts)
```

Step 3.
```{r}
treated.inds <- metadata$dex == "treated"
treated.counts <- counts[ , treated.inds]
```

Step 4.
```{r}
treated.mean <- rowMeans(treated.counts)
```

For ease of book-keeping we can store these together in one data frame called `meancounts`

```{r}
meancounts <- data.frame(control.mean, treated.mean)
head(meancounts)
```

```{r}
plot(meancounts)
```
This is screaming at me to log transform the data
```{r}
plot(meancounts, log="xy")
```
We use log2"fold-change" as a way to compare

```{r}
# treated/control
log2(10/10)
log2(20/10)
log2(10/20)

log2(40/10)
```
```{r}
meancounts$log2fc <- log2(meancounts$treated.mean/meancounts$control.mean)
head(meancounts)
```

A common "rule-of-thumb" threshold for calling something "up" regulated is a log2-fold-change of +2 or greater, and -2 or less for "down" regulated.

> Q. How many genes are "up" regulated at the +2 log2FC threshold?

> Q. How many genes are "down" regulated (at the -2 log2FC threshold)?


Filter out non-zero values

```{r}
y <- data.frame(a=c(1,5,0,5), b=c(1,0,5,5))
y
```
```{r}
which(y==0, arr.ind=TRUE)
```

```{r}
zero.inds <- which(meancounts[, 1:2] == 0, arr.ind = T)[,1]
mygenes <- meancounts[-zero.inds,]
```

> Q. How many genes are "up" regulated at the +2 log2FC threshold?

```{r}
sum(mygenes$log2fc >= 2)
```

> Q. How many genes are "down" regulated (at the -2 log2FC threshold)?

```{r}
sum(mygenes$log2fc <= -2)
```

## DESeq Analysis

Let's do this with DESeq2 and put some stats behind these numbers.

```{r, message=FALSE}
library(DESeq2)
```

DESeq wants 3 things for analysis, countData, colData, and design.

```{r}
dds <- DESeqDataSetFromMatrix(countData = counts,
                       colData = metadata,
                       design = ~dex)
```

The main function in the DESeq package to run analysis is called `DESeq()`

```{r}
dds <- DESeq(dds)
```

Get the results out of this DESeq object with the function `results()`

```{r}
res <- results(dds)
head(res)
```

## Volcano Plot

This is a plot of log2FC vs adjusted p-value

```{r}
plot(res$log2FoldChange, -log(res$padj))
abline(v=c(-2,2), col="red")
abline(h=-log(0.05), col="red")
```

## Save our result
```{r}
write.csv(res, file="myresults.csv")
```

## A nicer ggplot volcano plot

```{r}
library(ggplot2)
mycols <- rep("gray", nrow(res))
mycols[ abs(res$log2FoldChange) > 2] <- "blue"
mycols[ res$padj >= 0.05] <- "gray"

ggplot(res) + aes(res$log2FoldChange, 
                  -log(res$padj)) + 
  geom_point(col=mycols)

```


## Add annotation data

We need to add gene symbols, gene names, and other database ids to make my results useful for further analysis.

```{r}
head(res)
```

We have ENSEMBL database ids in our `res` object

```{r}
head( rownames(res) )
```

We can use the `mapIds()` function from bioconductor to help us

```{r}
library("AnnotationDbi")
library("org.Hs.eg.db")
```

Let's see what database id formats we can translate between

```{r}
columns(org.Hs.eg.db)
```

```{r}
res$symbol <- mapIds(org.Hs.eg.db,
                     keys=row.names(res), # Our genenames
                     keytype="ENSEMBL",        # The format of our genenames
                     column="SYMBOL",          # The new format we want to add
                     multiVals="first")
```

```{r}
head(res$symbol)
```

Add `GENENAME` then `ENTREZID`

```{r}
res$genename <- mapIds(org.Hs.eg.db,
                     keys=row.names(res), # Our genenames
                     keytype="ENSEMBL",        # The format of our genenames
                     column="GENENAME",          # The new format we want to add
                     multiVals="first")
```
```{r}
head(res$genename)
```

```{r}
res$entrez <- mapIds(org.Hs.eg.db,
                     keys=row.names(res), # Our genenames
                     keytype="ENSEMBL",        # The format of our genenames
                     column="ENTREZID",          # The new format we want to add
                     multiVals="first")
```
```{r}
head(res$entrez)
```

## Save my annotated results

```{r}
write.csv(res, file="myresults_annotated.csv")
```

## Pathway analysis

We will use the **gage** function from bioconductor package.

```{r}
library(pathview)
library(gage)
library(gageData)

data(kegg.sets.hs)

# Examine the first 2 pathways in this kegg set for humans
head(kegg.sets.hs, 2)
```
What **gage** wants as input is a named vector of importance i.e. a vector with labeled fold-changes.

```{r}
foldchanges <- res$log2FoldChange
names(foldchanges) <- res$entrez
head(foldchanges)
```

Run gage analysis:
```{r}
data(kegg.sets.hs)
keggres = gage(foldchanges, gsets=kegg.sets.hs)
```

What is in the results:
```{r}
attributes(keggres)
```
```{r}
head(keggres$less, 5)
```
Let's look at just one of these: hsa05310

```{r}
library(pathview)

pathview(gene.data=foldchanges, pathway.id="hsa05310")
```

Insert figure for this pathway

![Asthma pathway from KEGG with my differentially expressed genes highlighted](hsa05310.pathview.png)


